zebROS_ws/src/controllers_2023/src/elevator_controller.cpp

```diff
@@ -0,0 +1,443 @@
+    {
+        ROS_ERROR("Could not find elevator_zeroing_timeout");
+        return false;
+    }
+
+    if (!readIntoScalar(controller_nh, "stage_2_height", stage_2_height))
+    {
+        ROS_ERROR("Could not find stage_2_height");
+        return false;
+    }
+
+    if (!readIntoScalar(controller_nh, "motion_magic_velocity", motion_magic_velocity_fast))
+    {
+        ROS_ERROR("Could not find motion_magic_velocity");
+        return false;
+    }
+
+
+    if (!readIntoScalar(controller_nh, "motion_magic_acceleration", motion_magic_acceleration_fast))
+    {
+        ROS_ERROR("Could not find motion_magic_acceleration");
+        return false;
+    }
+
+
+    if (!readIntoScalar(controller_nh, "motion_s_curve_strength", motion_s_curve_strength))
+    {
+    	ROS_ERROR("Could not find motion_s_curve_strength");
+    	return false;
+    }
+
+
+    //get config values for the elevator talon
+    XmlRpc::XmlRpcValue elevator_params;
+    if (!controller_nh.getParam("elevator_joint", elevator_params))
+    {
+        ROS_ERROR("Could not find elevator_joint");
+        return false;
+    }
+
+    if (!controller_nh.getParam("MAX_HEIGHT_VAL", MAX_HEIGHT_VAL))
+    {
+        ROS_ERROR("Could not find MAX_HEIGHT_VAL");
+        return false;
+    
+    }
+
+    //initialize the elevator joint
+    if (!elevator_joint_.initWithNode(talon_command_iface, nullptr, controller_nh, elevator_params))
+    {
+        ROS_ERROR("Cannot initialize elevator joint!");
+    }
+
+    elevator_service_ = controller_nh.advertiseService("elevator_service", &ElevatorController_2023::cmdService, this);
+
+
+
+    return true;
+}
+
+void ElevatorController_2023::starting(const ros::Time &time)
+{
+    zeroed_ = false;
+    last_zeroed_  = false;
+    last_mode_ = hardware_interface::TalonMode_Disabled;
+    last_position_ = -1; // give nonsense position to force update on first time through update()
+    position_command_.writeFromNonRT(ElevatorCommand_2023());
+}
+
+void ElevatorController_2023::update(const ros::Time &time, const ros::Duration &/*duration*/)
+{
+    // If we hit the limit switch, (re)zero the position.
+    if (elevator_joint_.getReverseLimitSwitch())
+    {
+        ROS_INFO_THROTTLE(2, "ElevatorController_2023 : hit limit switch");
+        if (!last_zeroed_)
+        {
+            zeroed_ = true;
+            last_zeroed_ = true;
+            elevator_joint_.setSelectedSensorPosition(0);
+            elevator_joint_.setDemand1Type(hardware_interface::DemandType_ArbitraryFeedForward);
+            elevator_joint_.setDemand1Value(arb_feed_forward_up_low);
+        }
+    }
+    else
+    {
+        last_zeroed_ = false;
+    }
+
+
+    if (zeroed_) // run normally, seeking to various positions
+    {
+        elevator_joint_.setMode(hardware_interface::TalonMode_MotionMagic);
+        if (((elevator_joint_.getMode()) == (hardware_interface::TalonMode_Disabled)) && ((last_mode_) == (hardware_interface::TalonMode_Disabled)))
```

Comment:
Remove redundant parens here -variable names don't need to be put in their own set of parens, just the == expressions